File: start-museum/java/src/main/java/workshop/CuratorSafety.java
package workshop;

import com.github.copilot.rpc.McpStdioServerConfig;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.PermissionRequest;
import com.github.copilot.rpc.PermissionRequestResult;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.LinkOption;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Pattern;

public final class CuratorSafety {
    public static final List<String> WIKIPEDIA_TOOLS = List.of(
            "wikipedia-search",
            "wikipedia-readArticle");
    public static final String EXHIBIT_FILE_NAME = "exhibit.html";

    private static final Set<String> WIKIPEDIA_TOOL_NAMES = Set.of(
            "search",
            "readArticle",
            "wikipedia-search",
            "wikipedia-readArticle");
    private static final String WIKIPEDIA_REJECTION =
            "This session allows only the scoped Wikipedia search and article tools.";
    private static final String EXHIBIT_WRITE_REJECTION =
            "This session allows writing only exhibit.html in the application working directory.";
    private static final Pattern SOURCES_HEADING = Pattern.compile("(?im)^##\\s+Sources\\s*$");
    private static final Pattern SOURCE_LINE = Pattern.compile("^\\s*-\\s*(.+?):\\s*(https://\\S+)\\s*$");

    private CuratorSafety() {
    }

    public static McpStdioServerConfig wikipediaServer() {
        // SessionConfig should place this config in a map entry keyed "wikipedia".
        return new McpStdioServerConfig()
                .setCommand("npx")
                .setArgs(List.of("-y", "wikipedia-mcp@1.0.3"))
                .setWorkingDirectory(Path.of("").toAbsolutePath().normalize().toString())
                .setTools(List.of("search", "readArticle"));
    }

    public static PermissionHandler wikipediaPermissionHandler() {
        return (request, ignored) -> CompletableFuture.completedFuture(
                isAllowedWikipediaRequest(request)
                        ? PermissionRequestResult.approveOnce()
                        : PermissionRequestResult.reject(WIKIPEDIA_REJECTION));
    }

    // Parses model-written links, not evidence that an article exists or was consulted.
    public static SourceExtraction extractSources(String content) {
        if (content == null || content.isBlank()) {
            return new SourceExtraction("", List.of());
        }

        var matcher = SOURCES_HEADING.matcher(content);
        int headingStart = -1;
        int headingEnd = -1;
        while (matcher.find()) {
            headingStart = matcher.start();
            headingEnd = matcher.end();
        }
        if (headingStart < 0) {
            return new SourceExtraction(content.strip(), List.of());
        }

        String body = content.substring(0, headingStart).strip();
        String sourcesText = content.substring(headingEnd);
        List<Source> sources = new ArrayList<>();
        sourcesText.lines().forEach(line -> {
            var sourceMatcher = SOURCE_LINE.matcher(line);
            if (sourceMatcher.matches()) {
                String title = sourceMatcher.group(1).trim();
                String url = sourceMatcher.group(2).trim();
                if (!title.isEmpty() && !url.isEmpty()) {
                    sources.add(new Source(title, url));
                }
            }
        });
        return new SourceExtraction(body, sources);
    }

    public static PermissionHandler exhibitWritePermission(Path workingDirectory) {
        Path applicationDirectory = workingDirectory.toAbsolutePath().normalize();
        return (request, ignored) -> CompletableFuture.completedFuture(
                request != null
                        && "write".equals(request.getKind())
                        && !Boolean.TRUE.equals(request.getManagedApprovalRequired())
                        && isExhibitWrite(request.getExtensionData(), applicationDirectory)
                        ? PermissionRequestResult.approveOnce()
                        : PermissionRequestResult.reject(EXHIBIT_WRITE_REJECTION));
    }

    private static boolean isAllowedWikipediaRequest(PermissionRequest request) {
        if (request == null || !"mcp".equals(request.getKind()) || request.getExtensionData() == null
                || Boolean.TRUE.equals(request.getManagedApprovalRequired())) {
            return false;
        }
        Map<String, Object> details = request.getExtensionData();
        return "wikipedia".equals(details.get("serverName"))
                && details.get("toolName") instanceof String toolName
                && WIKIPEDIA_TOOL_NAMES.contains(toolName);
    }

    private static boolean isExhibitWrite(Map<String, Object> request, Path workingDirectory) {
        if (request == null || !(request.get("fileName") instanceof String fileName) || fileName.isBlank()
                || (request.containsKey("requestSandboxBypass")
                    && !Boolean.FALSE.equals(request.get("requestSandboxBypass")))) {
            return false;
        }
        try {
            Path candidate = workingDirectory.resolve(Path.of(fileName)).normalize();
            Path allowed = workingDirectory.resolve(EXHIBIT_FILE_NAME);
            return candidate.equals(allowed) && !Files.isSymbolicLink(allowed)
                    && !Files.isDirectory(allowed, LinkOption.NOFOLLOW_LINKS);
        } catch (InvalidPathException exception) {
            return false;
        }
    }

    public static final class ArtifactState {
        private final Path path;
        private final byte[] hash;

        private ArtifactState(Path path, byte[] hash) {
            this.path = path;
            this.hash = hash;
        }
    }

    public static ArtifactState captureArtifactState(Path workingDirectory, String fileName) throws IOException {
        if (fileName == null || fileName.isBlank() || fileName.equals(".") || fileName.equals("..")
                || fileName.contains("/") || fileName.contains("\\") || Path.of(fileName).isAbsolute()) {
            throw new IllegalArgumentException("Specify a single file name in the working directory.");
        }
        if (!Files.isDirectory(workingDirectory)) {
            throw new IOException("The artifact working directory must exist.");
        }
        Path path = workingDirectory.toAbsolutePath().normalize().resolve(fileName);
        return new ArtifactState(path, readArtifactHash(path, false));
    }

    public static void verifyArtifactUpdate(ArtifactState snapshot) throws IOException {
        byte[] hash = readArtifactHash(snapshot.path, true);
        if (hash == null || Arrays.equals(hash, snapshot.hash)) {
            throw new IOException("No output update was verified.");
        }
    }

    private static byte[] readArtifactHash(Path path, boolean requireNonempty) throws IOException {
        BasicFileAttributes attributes;
        try {
            attributes = Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
        } catch (NoSuchFileException exception) {
            return null;
        }
        if (!attributes.isRegularFile() || attributes.isSymbolicLink()
                || (requireNonempty && attributes.size() == 0)) {
            throw new IOException("No output update was verified. The target must be a nonempty regular nonsymlink file.");
        }
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            try (var input = new DigestInputStream(
                    Files.newInputStream(path, StandardOpenOption.READ, LinkOption.NOFOLLOW_LINKS), digest)) {
                input.transferTo(java.io.OutputStream.nullOutputStream());
            }
            return digest.digest();
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is unavailable.", exception);
        }
    }

    public record Source(String title, String url) {
    }

    public record SourceExtraction(String body, List<Source> sources) {
        public SourceExtraction {
            sources = List.copyOf(sources);
        }
    }
}

File: finished/dotnet/museum-exhibit-studio/Helpers/CuratorSafety.cs
using System.Security.Cryptography;
using GitHub.Copilot;
using GitHub.Copilot.Rpc;

namespace MuseumExhibitStudio.Helpers;

#pragma warning disable GHCP001 // Custom permission decisions are evaluation-only in SDK 1.0.11.

public sealed record ResearchSource(string Title, string Url);

public sealed record ExtractedSources(string Body, IReadOnlyList<ResearchSource> Sources);

public static class CuratorSafety
{
    public const string ExhibitFileName = "exhibit.html";

    public static IReadOnlyList<string> WikipediaTools { get; } =
    [
        "wikipedia-search",
        "wikipedia-readArticle"
    ];

    private static readonly HashSet<string> AllowedWikipediaToolNames =
    [
        "search",
        "readArticle",
        "wikipedia-search",
        "wikipedia-readArticle"
    ];

    // Return this config as the "wikipedia" value in SessionConfig.McpServers.
    public static McpStdioServerConfig WikipediaServer() => new()
    {
        Command = "npx",
        Args = ["-y", "wikipedia-mcp@1.0.3"],
        WorkingDirectory = Directory.GetCurrentDirectory(),
        Tools = ["search", "readArticle"]
    };

    public static Func<PermissionRequest, PermissionInvocation, Task<PermissionDecision>> WikipediaPermissionHandler() =>
        (request, _) =>
        {
            var decision = request is PermissionRequestMcp { ServerName: "wikipedia", ManagedApprovalRequired: not true } wikipedia &&
                           AllowedWikipediaToolNames.Contains(wikipedia.ToolName)
                ? PermissionDecision.ApproveOnce()
                : PermissionDecision.Reject(
                    "This session allows only the scoped Wikipedia search and article tools.");

            return Task.FromResult(decision);
        };

    // Parses model-written links, not evidence that an article exists or was consulted.
    public static ExtractedSources ExtractSources(string content)
    {
        if (string.IsNullOrEmpty(content))
        {
            return new ExtractedSources(string.Empty, []);
        }

        var lines = content.ReplaceLineEndings("\n").Split('\n');
        var sourcesIndex = Array.FindLastIndex(
            lines,
            line => line.Trim().Equals("## Sources", StringComparison.OrdinalIgnoreCase));

        if (sourcesIndex < 0)
        {
            return new ExtractedSources(content.Trim(), []);
        }

        var body = string.Join('\n', lines[..sourcesIndex]).Trim();
        var sources = new List<ResearchSource>();

        foreach (var line in lines[(sourcesIndex + 1)..])
        {
            var trimmed = line.Trim();
            if (!trimmed.StartsWith("- ", StringComparison.Ordinal))
            {
                continue;
            }

            var item = trimmed[2..].Trim();
            var separatorIndex = item.IndexOf(": https://", StringComparison.OrdinalIgnoreCase);
            if (separatorIndex <= 0)
            {
                continue;
            }

            var title = item[..separatorIndex].Trim();
            var url = item[(separatorIndex + 2)..].Trim();
            if (title.Length == 0 || !url.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            sources.Add(new ResearchSource(title, url));
        }

        return new ExtractedSources(body, sources.AsReadOnly());
    }

    public static Func<PermissionRequest, PermissionInvocation, Task<PermissionDecision>> ExhibitWritePermission(
        string workingDirectory)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(workingDirectory);

        var normalizedWorkingDirectory = Path.GetFullPath(workingDirectory);
        var allowedPath = Path.GetFullPath(Path.Combine(normalizedWorkingDirectory, ExhibitFileName));

        return (request, _) =>
        {
            var decision = request is PermissionRequestWrite { ManagedApprovalRequired: not true, RequestSandboxBypass: not true } write &&
                           IsAllowedExhibitPath(write.FileName, normalizedWorkingDirectory, allowedPath)
                ? PermissionDecision.ApproveOnce()
                : PermissionDecision.Reject(
                    "This session allows writing only exhibit.html in the application working directory.");

            return Task.FromResult(decision);
        };
    }

    private static bool IsAllowedExhibitPath(
        string fileName,
        string workingDirectory,
        string allowedPath)
    {
        if (string.IsNullOrWhiteSpace(fileName))
        {
            return false;
        }

        try
        {
            var requestedPath = Path.GetFullPath(fileName, workingDirectory);
            return requestedPath.Equals(allowedPath, OperatingSystem.IsWindows()
                       ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal) &&
                   new FileInfo(allowedPath).LinkTarget is null &&
                   !Directory.Exists(allowedPath);
        }
        catch (ArgumentException)
        {
            return false;
        }
    }

    public sealed class ArtifactState
    {
        internal string FullPath { get; }
        internal byte[]? Hash { get; }

        internal ArtifactState(string fullPath, byte[]? hash)
        {
            FullPath = fullPath;
            Hash = hash;
        }
    }

    public static ArtifactState CaptureArtifactState(string workingDirectory, string fileName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(workingDirectory);
        ArgumentException.ThrowIfNullOrWhiteSpace(fileName);
        if (fileName is "." or ".." || fileName.IndexOfAny(['/', '\\']) >= 0 || Path.IsPathRooted(fileName))
        {
            throw new ArgumentException("Specify a single file name in the working directory.", nameof(fileName));
        }
        if (!Directory.Exists(workingDirectory))
        {
            throw new DirectoryNotFoundException("The artifact working directory must exist.");
        }

        var fullPath = Path.GetFullPath(Path.Combine(workingDirectory, fileName));
        return new ArtifactState(fullPath, ReadArtifactHash(fullPath, requireNonempty: false));
    }

    public static void VerifyArtifactUpdate(ArtifactState snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        var hash = ReadArtifactHash(snapshot.FullPath, requireNonempty: true);
        if (hash is null || (snapshot.Hash is not null && hash.AsSpan().SequenceEqual(snapshot.Hash)))
        {
            throw new InvalidOperationException("No output update was verified.");
        }
    }

    private static byte[]? ReadArtifactHash(string path, bool requireNonempty)
    {
        FileAttributes attributes;
        try
        {
            attributes = File.GetAttributes(path);
        }
        catch (FileNotFoundException)
        {
            return null;
        }

        if ((attributes & (FileAttributes.Directory | FileAttributes.ReparsePoint | FileAttributes.Device)) != 0)
        {
            throw new InvalidOperationException("No output update was verified. The target must be a regular nonsymlink file.");
        }

        using var stream = File.OpenRead(path);
        if (!stream.CanSeek || (requireNonempty && stream.Length == 0))
        {
            throw new InvalidOperationException("No output update was verified. The target must be a nonempty regular file.");
        }
        return SHA256.HashData(stream);
    }
}

#pragma warning restore GHCP001
